File: Ontology/IClassExpressionVisitor.cs
namespace Ontology
{
    public interface IClassExpressionVisitor
    {
        void Visit(IClass                  @class                );
        void Visit(IObjectIntersectionOf   objectIntersectionOf  );
        void Visit(IObjectUnionOf          objectUnionOf         );
        void Visit(IObjectComplementOf     objectComplementOf    );
        void Visit(IObjectOneOf            objectOneOf           );
        void Visit(IObjectSomeValuesFrom   objectSomeValuesFrom  );
        void Visit(IObjectAllValuesFrom    objectAllValuesFrom   );
        void Visit(IObjectHasValue         objectHasValue        );
        void Visit(IObjectHasSelf          objectHasSelf         );
        void Visit(IObjectMinCardinality   objectMinCardinality  );
        void Visit(IObjectMaxCardinality   objectMinCardinality  );
        void Visit(IObjectExactCardinality objectExactCardinality);
        void Visit(IDataSomeValuesFrom     dataSomeValuesFrom    );
        void Visit(IDataAllValuesFrom      dataAllValuesFrom     );
        void Visit(IDataHasValue           dataHasValue          );
        void Visit(IDataMinCardinality     dataMinCardinality    );
        void Visit(IDataMaxCardinality     dataMaxCardinality    );
        void Visit(IDataExactCardinality   dataExactCardinality  );
    }

    public class ClassExpressionVisitor: IClassExpressionVisitor
    {
        public virtual void Visit(IClass                  @class                ) {}
        public virtual void Visit(IObjectIntersectionOf   objectIntersectionOf  ) {}
        public virtual void Visit(IObjectUnionOf          objectUnionOf         ) {}
        public virtual void Visit(IObjectComplementOf     objectComplementOf    ) {}
        public virtual void Visit(IObjectOneOf            objectOneOf           ) {}
        public virtual void Visit(IObjectSomeValuesFrom   objectSomeValuesFrom  ) {}
        public virtual void Visit(IObjectAllValuesFrom    objectAllValuesFrom   ) {}
        public virtual void Visit(IObjectHasValue         objectHasValue        ) {}
        public virtual void Visit(IObjectHasSelf          objectHasSelf         ) {}
        public virtual void Visit(IObjectMinCardinality   objectMinCardinality  ) {}
        public virtual void Visit(IObjectMaxCardinality   objectMaxCardinality  ) {}
        public virtual void Visit(IObjectExactCardinality objectExactCardinality) {}
        public virtual void Visit(IDataSomeValuesFrom     dataSomeValuesFrom    ) {}
        public virtual void Visit(IDataAllValuesFrom      dataAllValuesFrom     ) {}
        public virtual void Visit(IDataHasValue           dataHasValue          ) {}
        public virtual void Visit(IDataMinCardinality     dataMinCardinality    ) {}
        public virtual void Visit(IDataMaxCardinality     dataMaxCardinality    ) {}
        public virtual void Visit(IDataExactCardinality   dataExactCardinality  ) {}
    }
}

File: Web/ClientApp/src/app/Ontology/IClassExpressionVisitor.ts
import { IClass } from "./IClass";
import { IDataAllValuesFrom } from "./IDataAllValuesFrom";
import { IDataExactCardinality, IDataMaxCardinality, IDataMinCardinality } from "./IDataCardinality";
import { IDataHasValue } from "./IDataHasValue";
import { IDataSomeValuesFrom } from "./IDataSomeValuesFrom";
import { IObjectAllValuesFrom } from "./IObjectAllValuesFrom";
import { IObjectExactCardinality, IObjectMaxCardinality, IObjectMinCardinality } from "./IObjectCardinality";
import { IObjectComplementOf } from "./IObjectComplementOf";
import { IObjectHasSelf } from "./IObjectHasSelf";
import { IObjectHasValue } from "./IObjectHasValue";
import { IObjectIntersectionOf } from "./IObjectIntersectionOf";
import { IObjectOneOf } from "./IObjectOneOf";
import { IObjectSomeValuesFrom } from "./IObjectSomeValuesFrom";
import { IObjectUnionOf } from "./IObjectUnionOf";

export interface IClassExpressionVisitor
{
    Class                 (class$                : IClass                 ): void;
    ObjectIntersectionOf  (objectIntersectionOf  : IObjectIntersectionOf  ): void;
    ObjectUnionOf         (objectUnionOf         : IObjectUnionOf         ): void;
    ObjectComplementOf    (objectComplementOf    : IObjectComplementOf    ): void;
    ObjectOneOf           (objectOneOf           : IObjectOneOf           ): void;
    ObjectSomeValuesFrom  (objectSomeValuesFrom  : IObjectSomeValuesFrom  ): void;
    ObjectAllValuesFrom   (objectAllValuesFrom   : IObjectAllValuesFrom   ): void;
    ObjectHasValue        (objectHasValue        : IObjectHasValue        ): void;
    ObjectHasSelf         (objectHasSelf         : IObjectHasSelf         ): void;
    ObjectMinCardinality  (objectMinCardinality  : IObjectMinCardinality  ): void;
    ObjectMaxCardinality  (objectMaxCardinality  : IObjectMaxCardinality  ): void;
    ObjectExactCardinality(objectExactCardinality: IObjectExactCardinality): void;
    DataSomeValuesFrom    (dataSomeValuesFrom    : IDataSomeValuesFrom    ): void;
    DataAllValuesFrom     (dataAllValuesFrom     : IDataAllValuesFrom     ): void;
    DataHasValue          (dataHasValue          : IDataHasValue          ): void;
    DataMinCardinality    (dataMinCardinality    : IDataMinCardinality    ): void;
    DataMaxCardinality    (dataMaxCardinality    : IDataMaxCardinality    ): void;
    DataExactCardinality  (dataExactCardinality  : IDataExactCardinality  ): void;
}

export class ClassExpressionVisitor implements IClassExpressionVisitor
{
    Class                 (class$                : IClass                 ): void {}
    ObjectIntersectionOf  (objectIntersectionOf  : IObjectIntersectionOf  ): void {}
    ObjectUnionOf         (objectUnionOf         : IObjectUnionOf         ): void {}
    ObjectComplementOf    (objectComplementOf    : IObjectComplementOf    ): void {}
    ObjectOneOf           (objectOneOf           : IObjectOneOf           ): void {}
    ObjectSomeValuesFrom  (objectSomeValuesFrom  : IObjectSomeValuesFrom  ): void {}
    ObjectAllValuesFrom   (objectAllValuesFrom   : IObjectAllValuesFrom   ): void {}
    ObjectHasValue        (objectHasValue        : IObjectHasValue        ): void {}
    ObjectHasSelf         (objectHasSelf         : IObjectHasSelf         ): void {}
    ObjectMinCardinality  (objectMinCardinality  : IObjectMinCardinality  ): void {}
    ObjectMaxCardinality  (objectMaxCardinality  : IObjectMaxCardinality  ): void {}
    ObjectExactCardinality(objectExactCardinality: IObjectExactCardinality): void {}
    DataSomeValuesFrom    (dataSomeValuesFrom    : IDataSomeValuesFrom    ): void {}
    DataAllValuesFrom     (dataAllValuesFrom     : IDataAllValuesFrom     ): void {}
    DataHasValue          (dataHasValue          : IDataHasValue          ): void {}
    DataMinCardinality    (dataMinCardinality    : IDataMinCardinality    ): void {}
    DataMaxCardinality    (dataMaxCardinality    : IDataMaxCardinality    ): void {}
    DataExactCardinality  (dataExactCardinality  : IDataExactCardinality  ): void {}
}
